File: bindings/swift/Sources/EasyDogeKM/EasyDogeKM.swift
import Foundation
@_exported import easydoge_km_ffi

public enum StoredWalletProtection: Sendable, Equatable {
    case noPrompt
    case deviceCredential
    case biometric
}

public enum StorageProtectionLevel: Sendable, Equatable {
    case hardwareBacked
    case osBacked
    case unsupported
}

public struct StoredWalletHandle: Sendable, Equatable, Hashable {
    public let id: String

    public init(id: String) {
        self.id = id
    }
}

public protocol WalletSecretStore: Sendable {
    func storeMnemonic(
        _ mnemonic: String,
        protection: StoredWalletProtection
    ) async throws -> StoredWalletHandle

    func exportMnemonic(
        handle: StoredWalletHandle,
        protection: StoredWalletProtection
    ) async throws -> String

    func protectionLevel(handle: StoredWalletHandle) async throws -> StorageProtectionLevel
}

public struct EasyDogeKM: Sendable {
    public init() {}

    public func generateMnemonic(options: MnemonicOptions = MnemonicOptions(language: .english, wordCount: 24)) throws -> GeneratedMnemonic {
        try easydoge_km_ffi.generateMnemonic(options: options)
    }

    public func validateMnemonic(
        phrase: String,
        language: Language = .english
    ) throws -> Bool {
        try easydoge_km_ffi.validateMnemonic(phrase: phrase, language: language)
    }

    public func mnemonicToSeedHex(
        phrase: String,
        passphrase: String? = nil,
        language: Language = .english
    ) throws -> String {
        try easydoge_km_ffi.mnemonicToSeedHex(phrase: phrase, passphrase: passphrase, language: language)
    }

    public func accountKeys(
        phrase: String,
        passphrase: String?,
        language: Language = .english,
        network: Network = .mainnet,
        account: UInt32 = 0
    ) throws -> AccountKeySet {
        try easydoge_km_ffi.accountXprivFromMnemonic(
            phrase: phrase,
            passphrase: passphrase,
            language: language,
            network: network,
            account: account
        )
    }

    public func derivePathFromXpriv(xpriv: Xpriv, path: String) throws -> Xpriv {
        try easydoge_km_ffi.derivePathFromXpriv(xpriv: xpriv, path: path)
    }

    public func derivePathFromXpub(xpub: Xpub, path: String) throws -> Xpub {
        try easydoge_km_ffi.derivePathFromXpub(xpub: xpub, path: path)
    }

    public func xpubFromXpriv(xpriv: Xpriv) throws -> Xpub {
        try easydoge_km_ffi.xpubFromXpriv(xpriv: xpriv)
    }

    public func deriveAddressFromXpriv(xpriv: Xpriv, path: String) throws -> PathAddress {
        try easydoge_km_ffi.deriveAddressFromXpriv(xpriv: xpriv, path: path)
    }

    public func deriveAddressFromXpub(xpub: Xpub, path: String) throws -> PathAddress {
        try easydoge_km_ffi.deriveAddressFromXpub(xpub: xpub, path: path)
    }

    public func inspectXpriv(xpriv: Xpriv) throws -> ExtendedKeyInfo {
        try easydoge_km_ffi.inspectXpriv(xpriv: xpriv)
    }

    public func inspectXpub(xpub: Xpub) throws -> ExtendedKeyInfo {
        try easydoge_km_ffi.inspectXpub(xpub: xpub)
    }

    public func wifFromXpriv(xpriv: Xpriv) throws -> String {
        try easydoge_km_ffi.wifFromXpriv(xpriv: xpriv)
    }

    public func addressFromWif(network: Network = .mainnet, wif: String) throws -> WifInfo {
        try easydoge_km_ffi.addressFromWif(network: network, wif: wif)
    }

    public func validateAddress(network: Network = .mainnet, address: String) throws -> Bool {
        try easydoge_km_ffi.validateAddress(network: network, address: address)
    }

    public func createMultisigDescriptor(
        network: Network = .mainnet,
        threshold: UInt8,
        cosignerXpubs: [Xpub],
        childPath: String,
        sorted: Bool = true
    ) throws -> MultisigDescriptor {
        try easydoge_km_ffi.createMultisigDescriptor(
            network: network,
            threshold: threshold,
            cosignerXpubs: cosignerXpubs,
            childPath: childPath,
            sorted: sorted
        )
    }

    public func signMessage(
        network: Network = .mainnet,
        wif: String,
        message: String
    ) throws -> MessageSignature {
        try easydoge_km_ffi.signMessage(network: network, wif: wif, message: message)
    }

    public func verifyMessage(
        network: Network = .mainnet,
        address: String,
        signatureBase64: String,
        message: String
    ) throws -> Bool {
        try easydoge_km_ffi.verifyMessage(
            network: network,
            address: address,
            signatureBase64: signatureBase64,
            message: message
        )
    }

    public func signP2pkhTransaction(
        network: Network = .mainnet,
        unsignedTxHex: String,
        inputIndex: UInt64,
        scriptPubkeyHex: String,
        wif: String,
        sighashType: UInt32 = 1
    ) throws -> SignedTransaction {
        try easydoge_km_ffi.signP2pkhTransaction(
            network: network,
            unsignedTxHex: unsignedTxHex,
            inputIndex: inputIndex,
            scriptPubkeyHex: scriptPubkeyHex,
            wif: wif,
            sighashType: sighashType
        )
    }

    public func signSigningEnvelope(envelope: SigningEnvelope, wif: String) throws -> SigningEnvelope {
        try easydoge_km_ffi.signSigningEnvelope(envelope: envelope, wif: wif)
    }

    public func combineSigningEnvelopes(envelopes: [SigningEnvelope]) throws -> SigningEnvelope {
        try easydoge_km_ffi.combineSigningEnvelopes(envelopes: envelopes)
    }

    public func finalizeSigningEnvelope(envelope: SigningEnvelope) throws -> SignedTransaction {
        try easydoge_km_ffi.finalizeSigningEnvelope(envelope: envelope)
    }

    public func composeAndSignTransaction(request: ComposeTransactionRequest) throws -> ComposeTransactionResult {
        try easydoge_km_ffi.composeAndSignTransaction(request: request)
    }
}

File: bindings/kotlin/easydoge-km/src/main/java/io/easydoge/km/EasyDogeKM.kt
package io.easydoge.km

import uniffi.easydoge_km_ffi.AccountKeySet
import uniffi.easydoge_km_ffi.ExtendedKeyInfo
import uniffi.easydoge_km_ffi.GeneratedMnemonic
import uniffi.easydoge_km_ffi.Language
import uniffi.easydoge_km_ffi.MessageSignature
import uniffi.easydoge_km_ffi.MnemonicOptions
import uniffi.easydoge_km_ffi.MultisigDescriptor
import uniffi.easydoge_km_ffi.Network
import uniffi.easydoge_km_ffi.PathAddress
import uniffi.easydoge_km_ffi.SignedTransaction
import uniffi.easydoge_km_ffi.SigningEnvelope
import uniffi.easydoge_km_ffi.WifInfo
import uniffi.easydoge_km_ffi.Xpriv
import uniffi.easydoge_km_ffi.Xpub

data class StoredWalletHandle(val id: String)

enum class StoredWalletProtection {
    NoPrompt,
    DeviceCredential,
    Biometric,
}

enum class StorageProtectionLevel {
    HardwareBacked,
    OsBacked,
    Unsupported,
}

interface WalletSecretStore {
    suspend fun storeMnemonic(
        mnemonic: String,
        protection: StoredWalletProtection,
    ): StoredWalletHandle

    suspend fun exportMnemonic(
        handle: StoredWalletHandle,
        protection: StoredWalletProtection,
    ): String

    suspend fun protectionLevel(handle: StoredWalletHandle): StorageProtectionLevel
}

class EasyDogeKM {
    fun generateMnemonic(options: MnemonicOptions = MnemonicOptions(Language.ENGLISH, 24u)): GeneratedMnemonic =
        uniffi.easydoge_km_ffi.generateMnemonic(options)

    fun validateMnemonic(
        phrase: String,
        language: Language = Language.ENGLISH,
    ): Boolean = uniffi.easydoge_km_ffi.validateMnemonic(phrase, language)

    fun mnemonicToSeedHex(
        phrase: String,
        passphrase: String? = null,
        language: Language = Language.ENGLISH,
    ): String = uniffi.easydoge_km_ffi.mnemonicToSeedHex(phrase, passphrase, language)

    fun accountKeys(
        phrase: String,
        passphrase: String?,
        language: Language = Language.ENGLISH,
        network: Network = Network.MAINNET,
        account: UInt = 0u,
    ): AccountKeySet =
        uniffi.easydoge_km_ffi.accountXprivFromMnemonic(phrase, passphrase, language, network, account)

    fun derivePathFromXpriv(
        xpriv: Xpriv,
        path: String,
    ): Xpriv = uniffi.easydoge_km_ffi.derivePathFromXpriv(xpriv, path)

    fun derivePathFromXpub(
        xpub: Xpub,
        path: String,
    ): Xpub = uniffi.easydoge_km_ffi.derivePathFromXpub(xpub, path)

    fun xpubFromXpriv(xpriv: Xpriv): Xpub = uniffi.easydoge_km_ffi.xpubFromXpriv(xpriv)

    fun deriveAddressFromXpriv(
        xpriv: Xpriv,
        path: String,
    ): PathAddress = uniffi.easydoge_km_ffi.deriveAddressFromXpriv(xpriv, path)

    fun deriveAddressFromXpub(
        xpub: Xpub,
        path: String,
    ): PathAddress = uniffi.easydoge_km_ffi.deriveAddressFromXpub(xpub, path)

    fun inspectXpriv(xpriv: Xpriv): ExtendedKeyInfo = uniffi.easydoge_km_ffi.inspectXpriv(xpriv)

    fun inspectXpub(xpub: Xpub): ExtendedKeyInfo = uniffi.easydoge_km_ffi.inspectXpub(xpub)

    fun wifFromXpriv(xpriv: Xpriv): String = uniffi.easydoge_km_ffi.wifFromXpriv(xpriv)

    fun addressFromWif(
        network: Network = Network.MAINNET,
        wif: String,
    ): WifInfo = uniffi.easydoge_km_ffi.addressFromWif(network, wif)

    fun validateAddress(
        network: Network = Network.MAINNET,
        address: String,
    ): Boolean = uniffi.easydoge_km_ffi.validateAddress(network, address)

    fun createMultisigDescriptor(
        network: Network = Network.MAINNET,
        threshold: UByte,
        cosignerXpubs: List<Xpub>,
        childPath: String,
        sorted: Boolean = true,
    ): MultisigDescriptor =
        uniffi.easydoge_km_ffi.createMultisigDescriptor(network, threshold, cosignerXpubs, childPath, sorted)

    fun signMessage(
        network: Network = Network.MAINNET,
        wif: String,
        message: String,
    ): MessageSignature = uniffi.easydoge_km_ffi.signMessage(network, wif, message)

    fun verifyMessage(
        network: Network = Network.MAINNET,
        address: String,
        signatureBase64: String,
        message: String,
    ): Boolean = uniffi.easydoge_km_ffi.verifyMessage(network, address, signatureBase64, message)

    fun signP2pkhTransaction(
        network: Network = Network.MAINNET,
        unsignedTxHex: String,
        inputIndex: ULong,
        scriptPubkeyHex: String,
        wif: String,
        sighashType: UInt = 1u,
    ): SignedTransaction =
        uniffi.easydoge_km_ffi.signP2pkhTransaction(
            network,
            unsignedTxHex,
            inputIndex,
            scriptPubkeyHex,
            wif,
            sighashType,
        )

    fun signSigningEnvelope(
        envelope: SigningEnvelope,
        wif: String,
    ): SigningEnvelope = uniffi.easydoge_km_ffi.signSigningEnvelope(envelope, wif)

    fun combineSigningEnvelopes(envelopes: List<SigningEnvelope>): SigningEnvelope =
        uniffi.easydoge_km_ffi.combineSigningEnvelopes(envelopes)

    fun finalizeSigningEnvelope(envelope: SigningEnvelope): SignedTransaction =
        uniffi.easydoge_km_ffi.finalizeSigningEnvelope(envelope)
}
